generator-jhipster-go/generators/server/templates/go/src/services/authservice.go

```diff
@@ -0,0 +1,217 @@
+// var client gocloak.GoCloak
+
+func InitializeOauthServer() {
+	// fmt.Print(provider.Endpoint())
+	// oauth2Config := oauth2.Config{
+	// 	ClientID:     clientId,
+	// 	ClientSecret: clientSecret,
+	// 	RedirectURL:  redirectURL,
+	// 	Endpoint: provider.Endpoint(),
+	// 	Scopes: []string{oidc.ScopeOpenID, "profile", "email"},
+	// }
+}
+
+func Protect(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		provider, err := oidc.NewProvider(r.Context(), configURL)
+        if err != nil {
+          panic(err)
+        }
+		fmt.Print(provider.Endpoint())
+	    oidcConfig := &oidc.Config{
+         ClientID: clientId,
+        }
+        verifier := provider.Verifier(oidcConfig)
+		authHeader := r.Header.Get("Authorization")
+		if len(authHeader) < 1 {
+			w.WriteHeader(401)
+			json.NewEncoder(w).Encode(errors.UnauthorizedError())
+			return
+		}
+	    // token, _ := client.LoginClient(r.Context(), clientId,clientSecret, realm)
+        // fmt.Println(token)
+		accessToken := strings.Split(authHeader, " ")[1]
+		fmt.Println(accessToken);
+		
+		_,err = verifier.Verify(r.Context(), accessToken)
+
+		if err != nil {
+			w.WriteHeader(400)
+			json.NewEncoder(w).Encode(errors.BadRequestError(err.Error()))
+			return
+		}
+
+		next.ServeHTTP(w, r)
+	})
+}
+
+type loginRequest struct {
+	Username string `json:"username"`
+	Password string `json:"password"`
+}
+
+type Result struct {
+	access_token  string `json:"access_token"`
+	token_type   string `json:"token_type"`
+	refresh_token string `json:"refresh_token"`
+	expiry    string    `json:"expiry"`
+}
+
+func Redirect(w http.ResponseWriter, r *http.Request){
+	provider, err := oidc.NewProvider(r.Context(), configURL)
+	if err != nil {
+		panic(err)
+	  }
+	//   fmt.Print(provider.Endpoint())
+	oidcConfig := &oidc.Config{
+		ClientID: clientId,
+	   }
+	verifier := provider.Verifier(oidcConfig)
+	oauth2Config := oauth2.Config{
+		ClientID:     clientId,
+		// ClientSecret: clientSecret,
+		RedirectURL:  redirectURL,
+		Endpoint: provider.Endpoint(),
+		Scopes: []string{oidc.ScopeOpenID, "profile", "email"},
+	}
+
+	if r.URL.Query().Get("state") != state {
+		http.Error(w, "state did not match", http.StatusBadRequest)
+		return
+	}
+
+	oauth2Token, err := oauth2Config.Exchange(r.Context(), r.URL.Query().Get("code"))
+	if err != nil {
+		http.Error(w, "Failed to exchange token: "+err.Error(), http.StatusInternalServerError)
+		return
+	}
+	// fmt.Println(r.URL.Query().Get("code"))
+	// atoken :=oauth2Token [0]
+	// fmt.Println(atoken)
+	// fmt.Println(oauth2Token)
+	rawIDToken, ok := oauth2Token.Extra("id_token").(string)
+	if !ok {
+		http.Error(w, "No id_token field in oauth2 token.", http.StatusInternalServerError)
+		return
+	}
+	idToken, err := verifier.Verify(r.Context(), rawIDToken)
+	if err != nil {
+		http.Error(w, "Failed to verify ID Token: "+err.Error(), http.StatusInternalServerError)
+		return
+	}
+	fmt.Println(rawIDToken)
```

Comment:
IDToken are not confidential, but if not required you can remove the printing in console